src/util/async_utils.py

```diff
@@ -201,3 +219,177 @@ async def __aenter__(self):
 
     async def __aexit__(self, *exc):
         pass
+
+
+def generate_async_methods_from(cls):
+    """Class decorator that generates asynchronous versions of methods using the provided class.
+
+    This class assumes that every method of the resulting class is either an async gen or async function. In case of
+    the naming collision (decorated class already has the method that would have been created by the decorator)
+    the new method is not generated. This allows you to custom implement asynchronous versions of non-trivial methods
+    while automatically generating boilerplate code.
+    """
+
+    def _substitute_for_loop(match):
+        method_name = match.group(2)
+        method = getattr(cls, method_name)
+        if inspect.isgeneratorfunction(method):
+            return match.expand(r'async for \1 in self.\2\3:')
+        else:
+            return match.expand(r'for \1 in await self.\2\3:')
+
+    def _substitute_method(match):
+        method_name = match.group(1)
+        method = getattr(cls, method_name)
+        if inspect.isgeneratorfunction(method):
+            return match.expand(r'self.\1(')
+        else:
+            return match.expand(r'await self.\1(')
+
+    def _generate_async_version_source(method):
+        yield_from_regex = re.compile(r'yield from (\S+)')
+        source = re.sub(
+            yield_from_regex,
+            r'async for _val in \1: yield _val',
+            inspect.getsource(method),
+        )
+
+        # Generator[YieldType, SendType, ReturnType] -> AsyncGenAdapter[YieldType, SendType]
+        generator_type_annotation_regex = re.compile(r'Generator\[(\S+, \S+), \S+]')
+        source = re.sub(generator_type_annotation_regex, r'AsyncGenAdapter[\1]', source)
+
+        # Generator.send -> AsyncGenerator.asend
+        send_regex = re.compile(r'(\w+)\.send\(')
+        source = re.sub(send_regex, r'await \1.asend(', source)
+
+        # method calls
+        method_call_regex = re.compile(r'(?<!in )self\.(\w+)\(')
+        source = re.sub(method_call_regex, _substitute_method, source)
+
+        # for loops
+        for_loop_regex = re.compile(r'for (\w+) in self\.(\w+)([^:]+):')
+        source = re.sub(for_loop_regex, _substitute_for_loop, source)
+
+        # method signatures
+        source = source.replace(
+            f'def {method.__name__}', f'async def {method.__name__}'
+        )
+
+        # headers
+        source = source.replace("add_headers('client')", "add_headers('async_client')")
+
```

Comment:
This will require that all context managers that are used/will be used in TolokaClient implement `__aenter__`, `__aexit__` interface. I would not enforce that: let's require a custom implementation of the async method if there is truly a need for async context manager usage